site-src/geps/gep-709.md

```diff
@@ -0,0 +1,260 @@
+    kind: HTTPRoute
+    namespace: foo
+  to:
+  - group: core
+    kind: Service
+```
+
+### API
+This proposed API is fairly straightforward, but comes with a couple notable
+decisions:
+
+1. Each ReferencePolicy only supports a single From and To section. Additional
+   trust relationships can be modeled with additional ReferencePolicy resources.
+1. Resource names are intentionally excluded from this policy for simplicity and
+   because they rarely provide any meaningful protection.
+1. A single Namespace is allowed per "From" struct. Although a selector would be
+   more powerful it may encourage unnecessarily insecure configuration.
+
+```go
+// ReferencePolicy identifies cross namespace relationships that are trusted for
+// Gateway API.
+type ReferencePolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of ReferencePolicy.
+    Spec ReferencePolicySpec `json:"spec,omitempty"`
+}
+
+
+// ReferencePolicySpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferencePolicySpec struct {
+    // From describes the trusted namespaces and kinds that can reference the
+    // resources described in "To".
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinItems=1
+    From []ReferencePolicyFrom `json:"from"`
+
+    // To describes the resources that may be referenced by the resources
+    // described in "From".
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinItems=1
+    To []ReferencePolicyTo `json:"to"`
+}
+
+// ReferencePolicyFrom describes trusted namespaces and kinds.
+type ReferencePolicyFrom struct {
+    // Group is the group of the referrent.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Group string `json:"group"`
+
+    // Kind is kind of the referrent.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Kind string `json:"kind"`
+
+    // Namespace is the namespace of the referrent.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Namespace string `json:"namespace,omitempty"`
+}
+
+// ReferencePolicyTo describes trusted kinds.
+type ReferencePolicyTo struct {
+    // Group is the group of the referrent.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Group string `json:"group"`
+
+    // Kind is kind of the referrent.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Kind string `json:"kind"`
+}
+```
+
+
+### Benefits
+
+* Conceptually similar to NetworkPolicy.
+* A separate resource enables admins to restrict who can allow cross namespace
```

Comment:
Can you give an example of when this would actually be useful? I am having a hard time thinking of a scenario where I would want to disallow a namespace from exposing themselves to other namespaces

Reply:
I think this is conceptually similar to organizations who want to restrict the namespaces that can create Services of type LB and use quotas to achieve that. It's possible that you want users to be able to create Services without allowing them to expose them externally.

Reply:
The use case for not allowing things to expose themselves to other namespaces is any cluster where a namespace admin for one namespace is not totally trusted. In that case, allowing inbound references is a critical step in privilege escalation (another reason to only allow this interaction in one direction).

For example, I've worked before on CI clusters, where each namespace ran a step in a CI pipeline. Now, CI pipelines are essentaily remote-code-execution-as-a-service, so those workloads are very untrusted. Would someone here be able to create a Service record? Possibly, and if so, it's almost certainly good to ensure that they cannot create cross-namespace references to their namespace.

This is a little contrived, I know, but I think the principle of "relatively untrusted namespace owner" is a pretty sound usecase for this overall.

Reply:
Hmm, I am not sure I understand. Can you let me know where I am going wrong? This is my understanding:

Lets say we have a namespace `ingress`, which contains a Gateway. We have a namespace `untrusted`, which has a Service.

I think the statement here is we are trying to avoid letting `untrusted` expose their service through `ingress`' LoadBalancer. We are NOT trying to restrict `untrusted` from creating an HTTPRoute though (that is independent of this and driven by Gateway/Route selection agreement); rather, we are trying to restrict `ingress` from creating an HTTPRoute referencing `untrusted`.

So by restricting `untrusted` from creating a ReferencePolicy, we _think_ we have achieved this. In reality, the `ingress` namespace can just point to something in its own namespace that forwards to `untrusted` (there are many ways to do this, for simplicity lets just say they set up a dumb TCP proxy that forwards all requests to `untrusted`). As a result, I would say that the use of restrictive RBAC policy on ReferencePolicy does not achieve the goal of not allowing a Service to expose itself.

Reply:
The point of the ReferencePolicy is to require the owner of the `untrusted` namespace to accept incoming references. If they can't create ReferencePolicy objects, then they can't accept incoming references, and so say, the admin can create a more-open HTTPRoute that may reference a lot of Services, but only allow some to complete the two-way handshake.

I agree that it's a bit convoluted, but the idea here is to put the control over incoming references into the hands of the person who wons the referent. The main advantage of doing this with a separate object is that it allows us to apply this pattern to things whose spec we can't change (like Service), while still keeping that control in the hands of the object's owner. I think that the RBAC advantage of having a separate object is a minor one at best compared to that, but it may prove useful to people who are extremely careful with their RBAC policies.

I probably overemphasised this usecase previously, sorry about that.

Reply:
> The point of the ReferencePolicy is to require the owner of the untrusted namespace to accept incoming references. If they can't create ReferencePolicy objects, then they can't accept incoming references, and so say, the admin can create a more-open HTTPRoute that may reference a lot of Services, but only allow some to complete the two-way handshake.

I think I understand what you are discussing now, but I don't think its a good idea. The fact its possible to do this with the current design is no problem - obviously we cannot stop users from doing obscure things, but I wouldn't want to recommend this.

Instead of the admin "handshaking" every namespace, but then restricting their ability to handshake back, it seems like they should just not "handshake" in the first place?

Reply:
Might be getting a bit side tracked here though :slightly_smiling_face:

Reply:
Agreed. I think you make a fair point, that it's a pretty unlikely scenario that we already provide betters tools to deal with. But I think that having the ability to RBAC the ReferencePolicy object separately is an advantage, just not that big of one.